### ISSUE TITLE
evtxdump UniDecodeError with MS DNS Server log file

### ISSUE THREAD

**Issue by stephensheridan** [williballenthin/python-evtx#28]
Firstly, thanks for writing this library. I'm having an issue with converting binary evtx files that are generated using Windows DNS server. I'm assuming that the error is being caused by a field containing maybe a DNS request that has characters that cannot be decoded to ascii. This would be pretty normal when dealing with DNS queries. See output below....

```
Traceback (most recent call last):
  File "evtxdump.py", line 53, in <module>
    main()
  File "evtxdump.py", line 48, in main
    for xml, record in evtx_file_xml_view(fh):
  File "build/bdist.macosx-10.12-intel/egg/Evtx/Views.py", line 221, in evtx_file_xml_view
  File "build/bdist.macosx-10.12-intel/egg/Evtx/Views.py", line 172, in _build_record_xml
  File "build/bdist.macosx-10.12-intel/egg/Evtx/Views.py", line 163, in rec
  File "build/bdist.macosx-10.12-intel/egg/Evtx/Views.py", line 156, in rec
  File "build/bdist.macosx-10.12-intel/egg/Evtx/BinaryParser.py", line 64, in __call__
  File "build/bdist.macosx-10.12-intel/egg/Evtx/Nodes.py", line 945, in fast_substitutions
  File "build/bdist.macosx-10.12-intel/egg/Evtx/BinaryParser.py", line 475, in unpack_string
UnicodeDecodeError: 'ascii' codec can't decode byte 0x96 in position 16: ordinal not in range(128)
```

**Comment by williballenthin** [comment]
thanks for reporting this issue! as you presumed, there's an issue parsing some string data. from what i can gather from the stacktrace, there's data declared as either ascii or utf-8 that doesn't get extracted correctly. i'm not sure if its bad data, or valid utf-8 data that's being shoveled into the ascii decoder.

unfortunately, due to the evtx structures, without an example file, i'll have a hard time preparing a fix. are you able to provide a test log file that i can inspect? if not, perhaps we can try to find an alternative, though it might take a bit longer.

**Comment by stephensheridan** [comment]
Hi, thank you for the reply. I understand it will be difficult to track down the problem without a sample log file. However, its a bit tricky for me to send you the log file I'm working on as it contains info on DNS lookups from with our organisation along with source IP etc. I might be able to look at the files using MS DNS server log file viewer in order to isolate the offending entry. If I can do this I could at least send on a smaller sample. I will see if I can work on this during the week and send on what I can.

I stumbled across your library because I need to analyse DNS log files and there is a bug in MS DNS server 2012/2016 that causes the exporting of evtx log fles as CSV or XML to fail. Our techs logged a call to MS and they said they are working on a fix but no timeline. I do most of my analysis using Python so your library is a great fit for my problem.

Regards
Stephen

**Comment by williballenthin** [comment]
ok, i completely understand. i certainly wouldn't want my logs shipped off to some guy on the internet ( ;-) ), so i don't expect you to. let me brainstorm a procedure for triaging this. if its affecting you, its probably affecting other people, so i'd really like to get this fixed up.

**Comment by stephensheridan** [comment]
Hi, 

I added a try block to my own code to see could I figure out whereabouts in the log file was causing the encoding error. I found the point where the xml conversion crashed and then used MS Event Viewer to seek to that point in order to look at the event contents. It looks like a bug in the code that generates the log file in the first place as the MS Event viewer is unable to read the problematic event, see attached screen grab of the Event Viewer.

I wonder is there a nice way to handle XML events that are NOT "well formed"?

I have isolated a small section of the evtx log file that I can share with you if you like?

![dns_log_problem](https://cloud.githubusercontent.com/assets/14109896/20679025/0fae8b96-b591-11e6-9769-6bd9c99010ce.png)

**Comment by williballenthin** [comment]
huh, this is tricky. there is a routine we can used to strip out the invalid string data; however, typically i avoid using it cause it tends to hide *my* bugs when dealing with ascii/unicode. here, it seems fair to use that function.

i'll implement and test a fix today.

if you're able, can you send me a fragment of that example file that contains the malformed string shown in the screenshot above? perhaps use a hex editor to extract from a few bytes (16? 32?) before the string, until a bunch of bytes afterwards (another 32 bytes?). feel free to redact sensitive bytes by replacing them with `A` or `X`.

**Comment by stephensheridan** [comment]
Hi, Sorry for the delay.

I extracted the offending records into a separate evtx file.

Tried to edit in hex editor to anonymise  but for some reason was corrupting file, tried to load back into MS event viewer and was getting corrupt file error.

So eventually gave up and exported offending sample as plain XML, anonymised and have attached. 

NOTE: I have renamed to .txt in order to attach
[dns-problem-log-xml.txt](https://github.com/williballenthin/python-evtx/files/631846/dns-problem-log-xml.txt)


You can see the offending characters in the QNAME field. I am not entirely sure why strange characters are appearing, code be a bug in the DNS server log generator or code well be strange encoding error.

Maybe you code revert attached file back to evtx format?

Regards
Stephen

**Comment by williballenthin** [comment]
i've just finished a substantial amount of work to refactor the encoding and handling of xml data in this library. i suspect this issue *should* be fixed, but without the source data, I cannot confirm. please try again using v0.6.0 onwards, and reopen the issue if anything crops up!

closed in 4b4ba9569c2dc1e7aa3d98d76ba92ce08d8ee0d7
